### PR TITLE
Add web export

### PR DESCRIPTION
**Pull request by dsnopek**
~~Marking as a DRAFT because this builds on the changes from PR #88 which should get reviewed and merged first~~

Some notes:

- In order to get around CORS on the Web, we need to add `&origin=*` to all the requests to the Wikipedia.
- ~~Since this project uses threads, it needs to have "Thread Support" enabled in the export settings, which requires some special headers from the web server. For this to work on Itch.io, you need to enable the "SharedArrayBuffer" option:~~
  Selection_278
  - ~~Note: We could probably rewrite it to not use threads on the web, and eliminate this constraint~~
  - **UPDATE:** In the current version of this PR, I've got non-threaded builds works so that we don't necessarily need this anymore. It works by spreading the time-consuming tasks across several frames. However, using a threaded build is still an option.
- For whatever reason, having 3D physics run on a separate thread would just crash on the web, so I disabled it. And also, for whatever reason, we can't override that option via a `.web` feature tag, like all the overrides I did on https://github.com/m4ym4y/museum-of-all-things/pull/37, so I had to completely disable it for all platforms. :-( Is this setting important for other platforms? Personally, I've never had good luck with running either physics or rendering on a separate thread - that's always caused weird problems for me
  - **UPDATE:** This setting also seems to cause a deadlock when exporting on GitHub Actions with Godot 4.4, so it's actually changed in PR #88 now, which this PR depends on

Anyway, it does work in my testing!

And if we can get the web working, we can also try to see if WebXR will work, which would be another way to distribute the VR version ;-)

**UPDATE (2025-04-02):** Here's some builds from this PR:

- Threads: https://static.snopekgames.com/cached/builds/moat-20250402-01-threads/
- No threads: https://static.snopekgames.com/cached/builds/moat-20250402-01-nothreads/

### CODE REVIEW

**Comment by dsnopek** [comment]
Nevermind, I got images working! So, this is only DRAFT until PR https://github.com/m4ym4y/museum-of-all-things/pull/37 is reviewed and merged

**Comment by m4ym4y** [comment]
That's really cool that this is possible! Thanks so much for getting this working!!

- As for the physics on a separate thread, I did notice it made a big performance improvement earlier in development but for some reason in the current version it doesn't seem to make much difference anymore. (It might have been before I moved a lot of the work into separate threads that I was noticing a benefit). I will test a little more to ensure that it doesn't make stuttering worse when loading exhibits (that was the main issue I had in the past because it seems like modifying a gridmap uses a fair bit of physics time)

- Does SharedArrayBuffer still have issues running on the web on mac? I could try removing the threading for the web platform but it's pretty deeply embedded in the game's logic. It does also help a lot with things that were otherwise causing large frame drops. I'm curious how some things on the web version are working, like the caching logic that uses FS operations, and the network calls

**Comment by dsnopek** [comment]
> I will test a little more to ensure that it doesn't make stuttering worse when loading exhibits

Cool, thanks, let me know if you encounter any issues!

> Does SharedArrayBuffer still have issues running on the web on mac?

Yes, there's still issues on Safari.

> I could try removing the threading for the web platform but it's pretty deeply embedded in the game's logic. It does also help a lot with things that were otherwise causing large frame drops.

Doing the web requests without threading should be easy on the web, because Godot just hands the request off to the browser which does its own threading, and gives us the result when it's done.

But, yeah, loading the image data and generating mipmaps would be challenging to do smoothly without threads - we'd have to make a queue where we do just a little bit every frame, or something like that.

I probably won't have time to mess with it for at least a week, but I may experiment with it later - it's an interesting problem :-)

> I'm curious how some things on the web version are working, like the caching logic that uses FS operations, and the network calls

On the web, Godot (via Emscripten) uses a fake file system on top of IndexDB - it works pretty transparently. And like I mentioned above, the network calls get handed off to the browser - they're done like normal AJAX requests from JavaScript.

**Comment by m4ym4y** [comment]
> On the web, Godot (via Emscripten) uses a fake file system on top of IndexDB - it works pretty transparently. And like I mentioned above, the network calls get handed off to the browser - they're done like normal AJAX requests from JavaScript.

that's an awesome feature!!

still probably worth removing the caching behavior for the web version though as the browser can manage its own cache for the images and articles pulled

also "data" settings can be disabled in the web version

**Comment by dsnopek** [comment]
> still probably worth removing the caching behavior for the web version though as the browser can manage its own cache for the images and articles pulled
> 
> also "data" settings can be disabled in the web version

Good point! When I can come back to this, I'll try disabling those things. (I'm doing a game jam for the next 7 days, so I won't have too much time until after)

**Comment by dsnopek** [comment]
A quick update on this PR:

- The current version removes all threads, and instead tries to pace work across several frames. This does mean that exhibits load slower, but on my desktop there is no hitching when things are being loaded, and the loading isn't _too_ slow, at least for me. I still need to test this on more browsers and platforms - the ultimate test will be how well it works on Safari :-)
- As requested, I've removed all the caching and disabled the "Data" tab in settings
- On Godot 4.3, there is a weird bug (only when using the single-threaded web build) where all 3D rendering would go black, either at startup or after going through 1 or 2 exhibits. Updating to Godot 4.4 seems to fix it, so this is rebased on PR https://github.com/m4ym4y/museum-of-all-things/pull/88, however, that's got at least one bug itself!

Anyway, due to all these changes, I've moved this back into DRAFT for now

**Comment by dsnopek** [comment]
I've now tested this on:

- **Linux with Chrome:** works fine!
- **Linux with Firefox:** works fine!
- **M1 Mac Mini with Safari:** mouse sensitivity needed to be cranked all the way up, but otherwise worked fine!
- **Chromebook with Chrome:** relatively low frame rate, and lots of noticeable hitching when exhibits are loading, but it's a very under-powered machine. I'm curious if a threaded version would perform better? I'll test it out when I get a chance. Otherwise, though, it does still basically work
- **Windows with Edge:** mostly works fine - it's a little hitchy here and there. This machine is much less powerful than my Linux box, which could be the reason. I may need to increase the frame pacing a bit for the texture queue, because most folks probably have a machine powered more like this one, than my Linux box

If anyone wants to test, here is a build of the current PR:

https://static.snopekgames.com/cached/builds/moat-20250312-01/

**Comment by dsnopek** [comment]
Here's a new build with some of the recent fixes from PR #88:

https://static.snopekgames.com/cached/builds/moat-20250314-01/

**Comment by dsnopek** [comment]
I tried splitting up the work that `DataManager` does to load an image into multiple queue items for time-consuming part. So, now there's a step for: request, load image, generate mipmaps and create texture.

This seems to have really helped on my ~8 year old Windows machine! There's barely any noticeable hitches anymore. My Chromebook still struggles pretty badly, but it feels like it might be a little better. Overall, though, I feel like this PR is in pretty good shape now. :-)

Here's a new build to test:

https://static.snopekgames.com/cached/builds/moat-20250315-01/

**Comment by dsnopek** [comment]
I've been doing a bunch of testing on different platforms with both a multi-threaded and single-threaded build, to see if one performs better or worse in each case.

For powerful machines (my main Linux box, and M1 Mac Mini), there is no difference with regard to hitches - both run hitch free. However, the threaded build will load exhibits _way_ faster.

For medium machines (my ~8 year old Windows box), there _may_ be a slight performance improvement with regard to hitches in the threaded build, but it's super hard to tell - they may actually be the same. Again, the exhibits load faster with threaded builds.

For the super low-end (my Chromebook), performance is bad and there's lots of hitches in both versions - there was no discernible difference.

Here's two builds if anyone else would like to do a similar comparison:

- Threads: https://static.snopekgames.com/cached/builds/moat-20250315-03-threads/
- No threads: https://static.snopekgames.com/cached/builds/moat-20250315-03-nothreads/

(These builds include the WebXR changes from #91, because I didn't want to make 4 builds, but I don't think it should matter for this kind of testing)

**Comment by dsnopek** [comment]
Now that PR https://github.com/m4ym4y/museum-of-all-things/pull/88 is merged, I'm taking this one out of DRAFT again!

**Comment by m4ym4y** [comment]
Awesome! I will test this out but I also think the mipmapping thread merged from #31 might require some accommodation here for the web version without threads

**Comment by dsnopek** [comment]
I've rebased and adapted the `MipmapThread` as described above! Here's some new builds (which also include the WebXR support):

- Threads: https://static.snopekgames.com/cached/builds/moat-20250401-01-threads/
- No threads: https://static.snopekgames.com/cached/builds/moat-20250401-01-nothreads/

**Comment by m4ym4y** [comment]
This is looking amazing!! For some reason though after the first time I loaded the application it started getting CORS issues when loading an exhibit. I am not sure why it worked at one point, but I think this could be because of the COMMON_HEADERS included in requestsync; I don't think you are allowed to set custom user agent from the browser without some CORS stuff. In the past web-based version I disabled the user-agent unless it was the desktop version.

Also, we may want to unbind crouch in this version, or perhaps put it on a key other than ctrl. Because crouching and walking forward closes the tab haha

> which also include the WebXR support

does this mean that this PR includes the content of the WebXR PR also?

anyways I'm super excited to have a web version available, is there anything you have left that needs to be done on this before I merge it?

**Comment by dsnopek** [comment]
> For some reason though after the first time I loaded the application it started getting CORS issues when loading an exhibit. I am not sure why it worked at one point, but I think this could be because of the COMMON_HEADERS included in requestsync; I don't think you are allowed to set custom user agent from the browser without some CORS stuff. In the past web-based version I disabled the user-agent unless it was the desktop version.

So, are you saying it's not working for you now, using one of the builds I linked to? If so, which browser and platform are you on?

I don't think setting custom headers should break anything, although, it may not actually work. CORS is mostly about the domain the web app is running on and a header (`Access-Control-Allow-Origin`) we get from the server we're making the AJAX request to (so, the wikipedia in this case). The CORS headers that we send should be automatic from the browser.

One thing to note is that it won't work if you try running from the filesystem (like with a `file:///` URL), you need to load it from a web server, but the web server built into Godot should work (ie clicking "Remote Debug" -> "Run in Browser" in the editor).

Anyway, to be safe, I've updated this PR to clear out the headers when making requests on the web.

> Also, we may want to unbind crouch in this version, or perhaps put it on a key other than ctrl. Because crouching and walking forward closes the tab haha

Hehe, yeah :-) Perhaps we could use SHIFT? Would it make sense to use SHIFT on all platforms so they are consistent?

For now, I've updated the PR to change the default to SHIFT only on the web.

> does this mean that this PR includes the content of the WebXR PR also?

No, sorry for the confusion. The build I linked to includes the changes from the WebXR PR because I was being lazy and didn't want to make 4 builds, but this PR only has the flat web changes.

Here's two new builds that only include the changes from this PR:

- Threads: https://static.snopekgames.com/cached/builds/moat-20250402-01-threads/
- No threads: https://static.snopekgames.com/cached/builds/moat-20250402-01-nothreads/

> anyways I'm super excited to have a web version available, is there anything you have left that needs to be done on this before I merge it?

Nope! Assuming it works in everyone's testing, this should be ready to go :-)

**Comment by dsnopek** [comment]
> Right now SHIFT is sprint, which is more useful than crouch. Maybe crouch could be 'c'? although it's a bit weird. We might just be able to not have crouch on the web version.

Ah, sorry, I didn't notice it was used for sprint. I guess I'll just unbind it when I make my next update.

> This was on Firefox on Linux, with both of the builds that you had linked

That is very strange! It works fine for me with Firefox 137 on Linux - in developer tools for me the `origin=*` is in there:

Selection_313

The only thing I can think is you somehow have an old version cached? I'm not sure how that could happen given that they all have different URLs, though. I'm really not sure what to suggest :-/

> What browser/OS are you using and do you not run into this issue?

I've just re-tested on Chrome and Firefox on Linux. However, in the past I've also tested Safari on MacOS and Edge on Windows

**Comment by m4ym4y** [comment]
Bizarre, I've tried hard reloads and disabling cache in the network tab but the issue persists. I'm going to try hosting the web version from my local version of the repo to rule out caching completely.

File: scenes/util/ThreadSafeWorkQueue.gd [hunk 1]
```diff
@@ -56,16 +62,31 @@ func add_item(name, item, _exhibit=null, front=false):
 func process_queue(name):
   var queue = _get_queue(name)
 
-  while true:
-    if _quitting:
-      return null
+  if Util.is_using_threads():
+    while true:
+      var item = _process_queue_item(queue)
+      if item:
```
Comment (m4ym4y) [inline scenes/util/ThreadSafeWorkQueue.gd:68]: if the item is null (as in when _quitting is true) this will not return, so the loop will continue even when quitting. This causes the game to hang on quit

**Comment by m4ym4y** [comment]
OK the web version hosted locally does not have any CORS issue. I guess it's just a mystery haha. I did note one more error but once that's fixed this looks good to merge! Thanks for debugging w me

File: scenes/util/ThreadSafeWorkQueue.gd (hunk 1 above)
Comment (dsnopek) [reply scenes/util/ThreadSafeWorkQueue.gd:68]: Ah, yeah, that should be `while not WorkQueue.get_quitting()` - I'll put that in the next update

Comment (dsnopek) [reply scenes/util/ThreadSafeWorkQueue.gd:68]: Fixed in my latest push!

**Comment by m4ym4y** [comment]
Alright looks like it's good to go!! I'm going to merge, thanks so much for making this feature! :)
